### PR TITLE
Add Periodic update for host devices

### PR DESCRIPTION
**Pull request by Mmduh-483**
Periodic update mechanism guarantees reporting the right number of
existing resouces

### CODE REVIEW

File: pkg/resources/server.go [hunk 1]
```diff
@@ -358,3 +368,45 @@ func (rs *resourceServer) NotifyRegistrationStatus(ctx context.Context, regstat
 	}
 	return &registerapi.RegistrationStatusResponse{}, nil
 }
+
+func (rs *resourceServer) UpdateDevices(devices []types.PciNetDevice) {
+	// Get device spec
+	deviceSpec := utils.GetDevicesSpec(devices)
+
+	// If not devices not changed skip
+	if !utils.DevicesChanged(rs.deviceSpec, deviceSpec) {
```
Comment (adrianchiris) [inline pkg/resources/server.go:377]: you can have `DevicesChanged` as non public method under resourceServer , or just in this file

File: pkg/resources/server.go [hunk 2]
```diff
@@ -104,21 +108,14 @@ func (rsc *resourcesServerPort) Dial(unixSocketPath string, timeout time.Duratio
...
 	}
 
-	for _, device := range devices {
-		rdmaDeviceSpec := device.GetRdmaSpec()
-		if len(rdmaDeviceSpec) == 0 {
-			log.Printf("Warning: non-Rdma Device %s\n", device.GetPciAddr())
-		}
-		deviceSpec = append(deviceSpec, rdmaDeviceSpec...)
-	}
+	deviceSpec := utils.GetDevicesSpec(devices)
```
Comment (adrianchiris) [inline pkg/resources/server.go:118]: why did you move this logic to utils  and not just keep it in this file ?

File: pkg/resources/server.go [hunk 3]
```diff
@@ -289,6 +290,15 @@ func (rs *resourceServer) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlu
 			// FIXME: there is no way to recover from the Unhealthy state.
 			d.Health = pluginapi.Unhealthy
 			_ = s.Send(&pluginapi.ListAndWatchResponse{Devices: rs.devs})
+		case <-rs.updateSignal:
+			log.Printf("Updating \"%s\" devices", rs.resourceName)
+			resp.Devices = rs.devs
+
+			if err := s.Send(resp); err != nil {
```
Comment (adrianchiris) [inline pkg/resources/server.go:297]: i dont think we want to fail here, just log and retry

File: pkg/types/types.go [hunk 4]
```diff
@@ -52,6 +53,7 @@ type ResourceManager interface {
 	StopAllServers() error
 	RestartAllServers() error
 	GetFilteredDevices(devices []PciNetDevice, selector *Selectors) []PciNetDevice
+	TriggerUpdate()
```
Comment (adrianchiris) [inline pkg/types/types.go:56]: just call this `PeriodicUpdate()` IMO its clearer

**Review by adrianchiris — CHANGES_REQUESTED** [review]
As always, Please add a descriptive commit message 

Explaining what functionality you want to add,
outlining the changes you did for each components

File: pkg/resources/resources_manager.go [hunk 5]
```diff
@@ -199,6 +205,8 @@ func (rm *resourceManager) DiscoverHostDevices() error {
 		log.Println("Warning: DiscoverHostDevices(): no PCI network device found")
 	}
 
+	rm.deviceList = []*ghw.PCIDevice{}
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/resources_manager.go:208]: To clean up the old list

File: pkg/resources/server.go (hunk 1 above)
Comment (Mmduh-483) [reply pkg/resources/server.go:377]: changed to private function in same file

File: pkg/resources/server.go (hunk 2 above)
Comment (Mmduh-483) [reply pkg/resources/server.go:118]: It is used in more than one location so moved to utils, do u prefer to do same as above

File: pkg/resources/server.go (hunk 3 above)
Comment (Mmduh-483) [reply pkg/resources/server.go:297]: changed

File: pkg/resources/server.go (hunk 2 above)
Comment (adrianchiris) [reply pkg/resources/server.go:118]: its still a method that is local to this file, so for now lets keep it here imo

File: pkg/resources/resources_manager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/resources/resources_manager.go:208]: can you add a comment above just to clarify this

e.g
// cleanup deviceList as this method is also called during periodic update for the resources

File: pkg/resources/resources_manager.go [hunk 6]
```diff
@@ -276,3 +284,24 @@ func (rm *resourceManager) GetFilteredDevices(devices []types.PciNetDevice,
 
 	return newDeviceList
 }
+
+func (rm *resourceManager) TriggerUpdate() {
+	if rm.PeriodicUpdateInterval > 0 {
+		go func() {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/resources_manager.go:290]: added

[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/resources/resources_manager.go:295]: and who stops this go routine ?

File: pkg/resources/server.go [hunk 7]
```diff
@@ -358,3 +366,75 @@ func (rs *resourceServer) NotifyRegistrationStatus(ctx context.Context, regstat
...
+	// Get device spec
+	deviceSpec := getDevicesSpec(devices)
+
+	// If not devices not changed skip
+	if !devicesChanged(rs.deviceSpec, deviceSpec) {
+		log.Printf("no changes to devices for \"%s\"", rs.resourceName)
+		log.Println("exposing devices: ", rs.devs)
+		return
+	}
+
```
Comment (adrianchiris) [inline pkg/resources/server.go:380]: the operation of updating `rs.deviceSpec` and `rs.devs` is not atomic so if an `allocate()` request is received while UpdateDevices occurs, it may not behave correctly.

I would add a mutex to resourceServer to protect access for both `rs.deviceSpec` and `rs.devs` then use it where these data structures are read/written

**Review by adrianchiris — CHANGES_REQUESTED** [review]
Added comments. Also according to commit message, the only changes that where performed are in resource manager. According to the actual code changes this is not the case.

can you update the commit message ?

File: pkg/resources/resources_manager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/resources_manager.go:295]: Changed to use time.ticker

**Review by adrianchiris — CHANGES_REQUESTED** [review]
you also need to protect Allocate as well

File: pkg/resources/resources_manager.go [hunk 8]
```diff
@@ -276,3 +285,40 @@ func (rm *resourceManager) GetFilteredDevices(devices []types.PciNetDevice,
...
+			}
+		}()
+
+		// Return stop function
+		return func() {
+			stopChan <- true
+			close(stopChan)
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/resources_manager.go:323]: changed

File: pkg/resources/server.go [hunk 9]
```diff
@@ -358,3 +368,79 @@ func (rs *resourceServer) NotifyRegistrationStatus(ctx context.Context, regstat
...
+	rs.deviceSpec = deviceSpec
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/server.go:387]: Changed

File: pkg/resources/server.go [hunk 10]
```diff
@@ -28,16 +29,19 @@ type resourcesServerPort struct {
...
+	socketPath     string
+	devs           []*pluginapi.Device
+	deviceSpec     []*pluginapi.DeviceSpec
+	stop           chan interface{}
+	stopWatcher    chan bool
+	updateResource chan bool
+	health         chan *pluginapi.Device
+	rsConnector    types.ResourceServerPort
+	rdmaHcaMax     int
+	mutex          sync.RWMutex
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/server.go:44]: added

File: pkg/resources/server.go [hunk 11]
```diff
@@ -289,6 +290,15 @@ func (rs *resourceServer) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlu
 			// FIXME: there is no way to recover from the Unhealthy state.
 			d.Health = pluginapi.Unhealthy
 			_ = s.Send(&pluginapi.ListAndWatchResponse{Devices: rs.devs})
+		case <-rs.updateResource:
+			log.Printf("Updating \"%s\" devices", rs.resourceName)
+			resp.Devices = rs.devs
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/server.go:295]: changed

File: pkg/resources/server.go [hunk 12]
```diff
@@ -279,7 +278,9 @@ func (rs *resourceServer) register() error {
 // ListAndWatch lists devices and update that list according to the health status
 func (rs *resourceServer) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
 	log.Println("exposing devices: ", rs.devs)
-	_ = s.Send(&pluginapi.ListAndWatchResponse{Devices: rs.devs})
+	resp := new(pluginapi.ListAndWatchResponse)
+	resp.Devices = rs.devs
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/server.go:282]: changed

File: pkg/resources/server.go [hunk 13]
```diff
@@ -28,16 +29,20 @@ type resourcesServerPort struct {
...
+	socketPath     string
+	devs           []*pluginapi.Device
+	deviceSpec     []*pluginapi.DeviceSpec
+	stop           chan interface{}
+	stopWatcher    chan bool
+	updateResource chan bool
+	health         chan *pluginapi.Device
+	rsConnector    types.ResourceServerPort
+	rdmaHcaMax     int
+	// For safe resources updating and prevent overlapping of resources before and after update
```
Comment (adrianchiris) [inline pkg/resources/server.go:44]: I dont understand the comment , which resources are you protecting ?

File: pkg/resources/server.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/resources/server.go:283]: you should either lock/unlock mutex or just post to the updateResource channel

[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/resources/server.go:284]: so adding a : `rs.UpdateResource <- true` did not work ?

**Review by adrianchiris — COMMENTED** [review]
Overall looks good Thanks!
lets add a commit for extending unit-tests with the changes introduced here.

File: pkg/resources/server.go [hunk 14]
```diff
@@ -28,16 +29,20 @@ type resourcesServerPort struct {
...
+	watchMode      bool
+	socketName     string
+	socketPath     string
+	stop           chan interface{}
+	stopWatcher    chan bool
+	updateResource chan bool
+	health         chan *pluginapi.Device
+	rsConnector    types.ResourceServerPort
+	rdmaHcaMax     int
+	// Mutex protects devs and dviceSpec
```
Comment (adrianchiris) [inline pkg/resources/server.go:42]: typo: dviceSpec -> deviceSpec

File: pkg/resources/server.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/server.go:284]: It worked after increasing the channel size, before it was failing with error:
```
goroutine 24 [chan send]:
```

File: pkg/resources/resources_manager_test.go [hunk 15]
```diff
@@ -485,4 +486,44 @@ var _ = Describe("ResourcesManger", func() {
...
+				configList:             configList,
+				resourceServers:        []types.ResourceServer{&fakeResourceServer},
+				netlinkManager:         nLink,
+				rds:                    rds,
+				PeriodicUpdateInterval: 1 * time.Millisecond,
+			}
+
+			rm.PeriodicUpdate()
+			time.Sleep(2 * time.Second)
+			Expect(len(rm.deviceList)).To(Equal(1))
```
Comment (adrianchiris) [inline pkg/resources/resources_manager_test.go:526]: can you also ensure that UpdateDevices was called on fakeResourceServer ?

File: pkg/resources/resources_manager_test.go [hunk 16]
```diff
@@ -485,4 +486,44 @@ var _ = Describe("ResourcesManger", func() {
...
+			rm := &resourceManager{
+				configList:             configList,
+				resourceServers:        []types.ResourceServer{&fakeResourceServer},
+				netlinkManager:         nLink,
+				rds:                    rds,
+				PeriodicUpdateInterval: 1 * time.Millisecond,
+			}
+
+			rm.PeriodicUpdate()
+			time.Sleep(2 * time.Second)
```
Comment (adrianchiris) [inline pkg/resources/resources_manager_test.go:525]: cleanup goroutine by calling its stop func ?

File: pkg/resources/resources_manager_test.go [hunk 17]
```diff
@@ -485,4 +486,44 @@ var _ = Describe("ResourcesManger", func() {
 			fakeResourceServer.AssertExpectations(testCallsAssertionReporter)
 		})
 	})
+	Context("PeriodicUpdate", func() {
+		It("Update resources for resource manager", func() {
```
Comment (adrianchiris) [inline pkg/resources/resources_manager_test.go:490]: can you also add a test when rm.PeriodicUpdateInterval is zero that no updates occur ?

**Review by adrianchiris — CHANGES_REQUESTED** [review]

File: pkg/resources/resources_manager_test.go (hunk 16 above)
Comment (Mmduh-483) [reply pkg/resources/resources_manager_test.go:525]: yes

File: pkg/resources/resources_manager_test.go (hunk 15 above)
Comment (Mmduh-483) [reply pkg/resources/resources_manager_test.go:526]: added

File: pkg/resources/resources_manager_test.go (hunk 17 above)
Comment (Mmduh-483) [reply pkg/resources/resources_manager_test.go:490]: added

File: pkg/resources/server.go [hunk 18]
```diff
@@ -358,3 +371,79 @@ func (rs *resourceServer) NotifyRegistrationStatus(ctx context.Context, regstat
...
+			devs = append(devs, dpDevice)
+		}
+		rs.devs = devs
+	}
+
+	rs.updateResource <- true
+}
+
+// devicesChanged detect if original and new devices are different
+func devicesChanged(deviceList, newDeviceList []*pluginapi.DeviceSpec) bool {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/resources/server.go:418]: added

**Review by adrianchiris — APPROVED** [review]
LGTM ! Thanks @Mmduh-483 

I have a feeling we may need a config knob for this periodic update.
can you open an issue to make `periodicUpdate` configured via CNI config

**Comment by adrianchiris** [comment]
will merge once CI passes
